### PR TITLE
Implement Configuration Generator

### PR DESCRIPTION
**Pull request by elmodaddyb**
This feature implements a separate generation mechanism for the Gerbera configuration system.  The list below outlines the changes.

### Fixes Issues
#209 - Config file example
#196 - Should not store UDN in config file

### New Features
* Add new command line option `--create-config` to generate configuration
* Move configuration generation to separate class implementation
* Generate UDN when generating `config.xml`

### Removed Features
* Remove automatic generation of `config.xml` during runtime
* Remove configuration migration capability

> This change affects how easily users can startup gerbera.  It puts ownership onto the user by removing automated generation of configuration.

All code was written using TDD so the functional coverage is adequate.  I ran the system through Valgrind and eliminated all possible memory leaks when running `--create-config`.  This is my first sizable implementation of C++.

Honest reviews welcome.  If there are incorrect c++ patterns please do let me know.

Thanks,

Eamonn

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ before_install:
 script:
   - mkdir build
   - cd build
-  - cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DWITH_SYSTEMD=0 -DWITH_TESTS=1 .. && make && make test
+  - cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DWITH_SYSTEMD=0 -DWITH_TESTS=1 .. && make && make test ARGS=-V
```
Comment (elmodaddyb) [inline .travis.yml]: This add verbose output of unit tests.  It is helpful when the test fails to know the exact failure.

File: config/config2.xsd [hunk 2]
```diff
@@ -149,11 +149,46 @@
     <xs:element name="extended-runtime-options">
         <xs:complexType>
             <xs:all>
+                <xs:element ref="ffmpegthumbnailer" minOccurs="0"/>
                 <xs:element ref="mark-played-items" minOccurs="0"/>
             </xs:all>
         </xs:complexType>
     </xs:element>
 
+    <xs:element name="ffmpegthumbnailer">
```
Comment (elmodaddyb) [inline config/config2.xsd]: When producing the fixtures to test, there was several missing XSD elements.  Cleanup and implement accurate XSD.

File: src/config/IConfigGenerator.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/// \file IConfigGenerator.h
+
+#ifndef GERBERA_ICONFIGGENERATOR_H
+#define GERBERA_ICONFIGGENERATOR_H
+
+class IConfigGenerator {
+
+ public:
+  virtual ~IConfigGenerator() {}
+  virtual std::string generate(std::string userHome, std::string configDir, std::string prefixDir, std::string magicFile) = 0;
```
Comment (elmodaddyb) [inline src/config/IConfigGenerator.h]: In the future we should eliminate `mxml` implementation and use a standard XML library.  This class begins to insulate the XML generation and expose standard signatures.

File: src/config/config_generator.cc [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+Ref<Element> ConfigGenerator::generateServer(std::string userHome, std::string configDir, std::string prefixDir) {
+  Ref<Element> server(new Element(_("server")));
+
+  server->appendElementChild(generateUi());
+  server->appendTextChild(_("name"), _(PACKAGE_NAME));
+
+  server->appendElementChild(generateUdn());
+
+  std::string homepath = userHome + DIR_SEPARATOR + configDir;
+  server->appendTextChild(_("home"), String::copy(homepath.c_str()));
```
Comment (elmodaddyb) [inline src/config/config_generator.cc]: This line had a memory leak.  By passing `std::string` into the function, but then returning the `Ref<Element>`.  The  **std::string** value is not used until later in the **generate** method.

I chose the `Ref<String>` implementation of `String::copy` which duplicates the string.  I previously tried using `strdup(...)` but this creates a `char*` which needs to be **free'd** later.

Any insight or suggestions here would be appreciated.

File: src/config/config_generator.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+#include "IConfigGenerator.h"
+
+class ConfigGenerator : public IConfigGenerator {
+ public:
+  ConfigGenerator();
+  ~ConfigGenerator();
+
+  std::string generate(std::string userHome, std::string configDir, std::string prefixDir, std::string magicFile) override;
+  zmm::Ref<mxml::Element> generateServer(std::string userHome, std::string configDir, std::string prefixDir);
```
Comment (elmodaddyb) [inline src/config/config_generator.h]: I looked for a suitable XML DOM generation in C++ but could not find anything easier to implement than the current `mxml::Element`.  Ideally this method will have standard library for XML and expose only the **generate** method.

File: src/config_manager.cc [hunk 6]
```diff
@@ -106,30 +107,18 @@ void ConfigManager::init()
...
 
     String home = userhome + DIR_SEPARATOR + config_dir;
-    bool home_ok = true;
 
     if (filename == nullptr) {
         // No config file path provided, so lets find one.
-        if (home_ok && (!check_path(userhome + DIR_SEPARATOR + config_dir + DIR_SEPARATOR + DEFAULT_CONFIG_NAME))) {
-            home_ok = false;
-        } else {
+        if (check_path(home + DIR_SEPARATOR + DEFAULT_CONFIG_NAME)) {
```
Comment (elmodaddyb) [inline src/config_manager.cc]: All we do now is:

1. Check if the default path is OK
2. Load default path OR throw exception

File: src/config_manager.cc [hunk 7]
```diff
@@ -106,30 +107,18 @@ void ConfigManager::init()
...
     }
 
-    if (filename == nullptr) {
-        throw _Exception(_("\nThe server configuration file could not be found in ~/.gerbera\n") + "Gerbera could not determine your home directory - automatic setup failed.\n" + "Try specifying an alternative configuration file on the command line.\n" + "For a list of options run: gerbera -h\n");
-    }
-
     log_info("Loading configuration from: %s\n", filename.c_str());
     load(filename);
-
-    prepare_udn();
```
Comment (elmodaddyb) [inline src/config_manager.cc]: Moved to `config_generator.cc`

File: src/main.cc [hunk 8]
```diff
@@ -236,20 +240,16 @@ For more information visit " DESC_MANUFACTURER_URL "\n\n");
...
+                create_config = true;
+                break;
             }
             break;
         default:
             break;
         }
     }
 
-    if (print_version) {
```
Comment (elmodaddyb) [inline src/main.cc]: Moved this **below** the `--create-config` execution.  This allows `--create-config` to generate cleanly.

File: test/test_config/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+add_custom_command(TARGET testconfig POST_BUILD
+        COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/fixtures $<TARGET_FILE_DIR:testconfig>/fixtures
+)
+
+add_test(NAME testconfig
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/test/test_config
+        COMMAND ./testconfig)
+
+add_definitions(-DCMAKE_BINARY_DIR="${CMAKE_BINARY_DIR}")
```
Comment (elmodaddyb) [inline test/test_config/CMakeLists.txt]: I added the `CMAKE_BINARY_DIR` to allow the tests to write files to the build/test directory.  This allows for more robust testing.

File: test/test_config/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        ${MYSQL_CLIENT_LIBS}
+        ${SQLITE3_LIBRARIES}
+        ${ICONV_LIBRARIES}
+        ${GTEST_LIBRARIES}
+        ${GERBERA_INTERFACE_LIBRARIES}
+        ${CMAKE_THREAD_LIBS_INIT}
+        )
+
+add_custom_command(TARGET testconfig POST_BUILD
+        COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/fixtures $<TARGET_FILE_DIR:testconfig>/fixtures
```
Comment (elmodaddyb) [inline test/test_config/CMakeLists.txt]: Copy all the **fixture** files to the build directory, thus allowing easy compare within GoogleTest

File: test/test_config/test_configgenerator.cc [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+};
+
+#if defined(HAVE_FFMPEG) && defined(HAVE_FFMPEGTHUMBNAILER) && defined(HAVE_MYSQL) && defined(HAVE_MAGIC) && defined(HAVE_JS) && defined(ONLINE_SERVICES)
+TEST_F(ConfigGeneratorTest, GeneratesFullConfigXmlWithAllDefinitions) {
+  std::string mockXml = mockConfigXml("fixtures/mock-config-all.xml");
+
+  std::string result = subject->generate(homePath, configDir, prefixDir, magicFile);
+
+  // remove UUID, for simple compare...TODO: mock UUID?
+  std::regex reg("<udn>uuid:[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}</udn>");
```
Comment (elmodaddyb) [inline test/test_config/test_configgenerator.cc]: Now that `udn` is generated, it is difficult to test.  I simply removed the value and assert `<udn/>`

The **udn** is tested independently.

File: test/test_config/test_configmanager.cc [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+class ConfigManagerTest : public ::testing::Test {
+
+ public:
+
+  ConfigManagerTest() {};
+
+  virtual ~ConfigManagerTest() {};
+
+  virtual void SetUp() {
+    gerberaDir = createTempPath();
```
Comment (elmodaddyb) [inline test/test_config/test_configmanager.cc]: `ConfigManager` requires a filesystem file to be initialized.  It also checks the validity of the script files referenced in config.

Initially was using **boost** library to generate temporary dir and files, but its a bit heavy weight to include just for testing.

File: test/test_config/test_configmanager.cc [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+  String config_file;
+  std::string home;
+  std::string confdir;
+  std::string prefix;
+  std::string magic;
+  ConfigManager *subject;
+};
+
+TEST_F(ConfigManagerTest, LoadsWebUIDefaultValues) {
+  subject->setStaticArgs(config_file, _(home.c_str()), _(confdir.c_str()), _(prefix.c_str()), _(magic.c_str()));
```
Comment (elmodaddyb) [inline test/test_config/test_configmanager.cc]: This test can be extended to test any of the configuration elements.  I plan to use the same to test the **SHOUTcast** capability.

File: test/test_server/test_main.cc [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  ServerTest() {};
+  virtual ~ServerTest() {};
+
+  virtual void SetUp() {
+  }
+
+  virtual void TearDown() {
+  };
+
+  std::string exec(const char* cmd) {
```
Comment (elmodaddyb) [inline test/test_server/test_main.cc]: Here we are actually launching the `gerbera` binary with the additional options to test its output.

**Comment by elmodaddyb** [comment]
When ready, we can **squash & merge** these revisions for simplicity.

**Comment by v00d00** [comment]
Looks good!

I do wonder if there is anything to gain anything by creating `IConfigGenerator` for a single method with a single implementation?

**Comment by elmodaddyb** [comment]
Yeah, I agree... I guess it was more an exercise in c++ learning for me. 

I will remove the abstract class `IConfigGenerator`

**Comment by elmodaddyb** [comment]
@v00d00 - I removed the `IConfigGenerator` abstract class.

Let me know if you have any other suggested improvements.

Thanks,

Eamonn

**Comment by v00d00** [comment]
Thanks! Merged, I squashed locally so that it kept your name on the commit.

**Pull request closed without merging**